File: mcts.py
import random
import math
import time

from game import Game
import numpy as np

class Node:
    C = 1
    def __init__(self, game, parent, player, move):
        self.game_state = game
        self.N = 0
        self.W = 0
        self.parent = parent
        self.move = move
        self.player = player
        self.children = [] # not sure it is useful

    def naive_rollout(self):
        topWon = False
        bottomWon = False
        player = self.player
        game = self.game_state
        while not (topWon or bottomWon):
            actions = game.get_valid_moves()
            action = random.choice(actions)
            game, topWon, bottomWon = game.apply_move(action)
            player = 1 - player
        
        self.N = 1
        self.W = 1 if player == 1 and topWon or player == 0 and bottomWon else -1

    def to_string(self, tabul):
        str = tabul + "Node G:{} A:{} N: {}, W: {}\n".format(self.game_state, self.move, self.N, self.W)
        for child in self.children:
            str+= child.to_string(tabul + "  ")
        return str
    
    def __str__(self):
        return self.to_string("")
    
    def back_propagate(self, n, w):
        self.W += w
        self.N += n
        if self.parent != None:
            self.parent.back_propagate(n, w)
    
    def expand(self):
        actions = self.game_state.get_valid_moves()
        if len(actions) == 0:
            return
        for a in actions:
            child_game,_,_ = self.game_state.apply_move(a)
            child_node = Node(child_game, self, 1 - self.player, a)
            child_node.naive_rollout()
            self.back_propagate(child_node.N, child_node.W)
            self.children.append(child_node)

    def select(self):
        if len(self.children) == 0:
            return self

        sign = -1 if self.player == 1 else 0

        UCT_scores = [sign * child.W / child.N + Node.C * math.sqrt(math.log(self.N)/child.N) for child in self.children]
        chosen_child_idx = np.argmax(UCT_scores)
        return self.children[chosen_child_idx].select()

    def run_mcts(node, time_budget=10):
        start_time = time.process_time()
        nb_selected_node = 0
        while time.process_time() - start_time < time_budget:
            selected_node = node.select()
            nb_selected_node += 1
            selected_node.expand()
        print("Selections:", nb_selected_node)
        # print(node)
        return max(node.children, key=lambda child: child.N).move
    
    def new_mcts_root(game):
        return Node(game, None, 0, None)

if __name__ == "__main__":
    game = Game.new()
    game.print()
    mcts = Node.new_mcts_root(game)
    move = Node.run_mcts(mcts)
    print("Choosen move:", move)

File: mcts.jl
using Profile

include("./schotten-game.jl")

const C = 1

mutable struct Node
    game_state::Schotten
    N::UInt
    W::Int
    parent::Nullable{Node}
    move::Tuple{Int64, Int64}
    player::Player
    children::Array{Node, 1}
    function Node(game::Schotten, parent::Nullable{Node}, player::Player, move:: Tuple{Int64, Int64})
        @assert player==top || player==bottom
        new(
            game,
            0,
            0,
            parent,
            move,
            player,
            []
        )
    end
end


function naive_rollout(node::Node, winner_already)
    winner = winner_already
    game = deepcopy(node.game_state)
    while winner == none
        actions = getvalidmoves(game)
        action = rand(actions)
        winner = applymove!(game, action)
    end

    
    node.N = 1
    node.W = node.player == winner ? 1 : -1
end

function show(io::IO, node::Node, tabul::String)
    str = tabul * "Node G:$node.game_state A:$game.move N: $game.N, W: $game.W"
    println(io, str)
    for child in node.children
        show(io, child, tabul * "  ")
    end
end

function Base.show(io::IO, node::Node)    
    show(io, node, "")
end

function back_propagate(node::Node, n::UInt, w::Int)
    node.W += w
    node.N += n
    if !isnull(node.parent)
        back_propagate(get(node.parent), n, w)
    end
end
    
function expand!(node::Node)
    actions = getvalidmoves(node.game_state)
    if length(actions) == 0
        return
    end
    for a in actions
        child_game = deepcopy(node.game_state)
        winner = applymove!(child_game, a)
        child_node = Node(child_game, Nullable{Node}(node), node.player, a)
        naive_rollout(child_node, winner)
        back_propagate(node, child_node.N, child_node.W)
        push!(node.children, child_node)
    end
end

function select(node)
    if length(node.children) == 0
        return node
    end
    sign = node.player == node.game_state.turn ? -1 : 1
    UCT_scores = [sign * child.W / child.N + C * sqrt(log(node.N)/child.N) for child in node.children]
    _, chosen_child_idx = findmax(UCT_scores)
    return select(node.children[chosen_child_idx])
end

function maxNode(nodes::AbstractArray{Node})
    len = length(nodes)
    @assert len != 0
    maxnode = nodes[1]
    if len > 1
        for node in nodes[2:end]
            if node.N > maxnode.N 
                maxnode = node
            end
        end
    end
    return maxnode
end

function run_mcts!(node, time_budget=10)
    start_time = time()
    nb_selected_node = 0
    while time() - start_time < time_budget
        selected_node = select(node)
        nb_selected_node += 1
        expand!(selected_node)
    end
    println("Selections:", nb_selected_node)
    # print(node)
    return maxNode(node.children).move
end
    
Node(game) = Node(game, Nullable{Node}(), top, (0,0))

function testperf(numiter, time_budget=10)
    for i in 1:numiter
        game = Schotten()
        # println(game)
        mcts = Node(game)
        move = run_mcts!(mcts, time_budget)
        println("Choosen move:", move)
    end
end

@time testperf(1,1)


Profile.clear()

@time testperf(1)


